rir/src/ir/Compiler.cpp

```diff
@@ -647,118 +661,315 @@ bool compileSpecialCall(CompilerContext& ctx, SEXP ast, SEXP fun, SEXP args_,
+             the RHS. See
+                https://github.com/reactorlabs/gnur/blob/R-3-6-2-branch-rir-patch/src/main/eval.c#L7128
+            */
+
+            bool const maybe_special = (fun2name == "$" || fun2name == "@");
+            if (maybe_special) {
+                return false;
+            }
 
-        if (maybeChanges(target, *idx) ||
-            (dims > 1 && maybeChanges(target, *(idx + 1))) ||
-            (dims > 2 && maybeChanges(target, *(idx + 2))))
-            cs << BC::setShared();
+            // Get the LISTSXP of args for f
+            SEXP f_args = CDR(CAR(args_));
+
+            // Make the ast for the call : f<-, x, y, value=z
+            // and protect it from GC
+            SEXP farrow_ast;
+            Protect farrow_ast_protect{farrow_ast =
+                                           Rf_lcons(farrow_sym, R_NilValue)};
+            // duplicate the args from the AST of the call to f into the AST for
+            // the call to `f<-` (directly linked in the AST so that everything
+            // is protected)
+            SETCDR(farrow_ast, Rf_duplicate(f_args));
+
+            SEXP last_farrow_cell = farrow_ast;
+            while (CDR(last_farrow_cell) != R_NilValue) {
+                last_farrow_cell = CDR(last_farrow_cell);
+            }
 
-        // And index
-        compileExpr(ctx, *idx);
-        if (dims > 1)
-            compileExpr(ctx, *(idx + 1));
-        if (dims > 2)
-            compileExpr(ctx, *(idx + 2));
+            // We need to append "value = z" to the list of args for f<-
+            // Let's create the corresponding cell (directly linked in AST so
+            // that it is protected)
+            SETCDR(last_farrow_cell, Rf_lcons(rhs, R_NilValue));
+            SEXP new_z_cell = CDR(last_farrow_cell);
+            SET_TAG(new_z_cell, Rf_install("value"));
+
+            // The RHS must be evaluated before the LHS
+            // Additionnaly, the value of the RHS must be returned after the
+            // assignment (in non-void contexts). It will be kept on the stack
+            // before the call to `f<-`.
+            // A copy will be wrapped in an evaluated promise and passed to f<-.
+            compileExpr(ctx, rhs);
 
-        if (dims == 3) {
-            assert(fun2 == symbol::Bracket);
-            cs << BC::subassign1_3();
-        } else if (dims == 2) {
-            if (fun2 == symbol::DoubleBracket) {
-                cs << BC::subassign2_2();
+            // Prepare the call to f<-(x, y1, <...>, yn, z)
+            cs << BC::ldfun(farrow_sym);
+
+            if (Compiler::profile)
+                cs << BC::recordCall();
+
+            // prepare x, yk, z as promises
+            LoadArgsResult load_arg_res;
+            SEXP farrow_args = CDR(farrow_ast);
+
+            // Load the value of x as an eager promise
+
+            compileLoadOneArg(ctx, farrow_args, ArgType::EAGER_PROMISE,
```

Comment:
the answer is probably "yes", but does `x` really need to be promise wrapped?

Reply:
This is not something I had tried before. I thought all arguments needed to be passed in promises: but if I try to pass the `x` and `z` (value) arguments as raw values, the (short) tests still pass